### PR TITLE
Fixes a possible N+1 issue within reclaim_space task

### LINKED ISSUE #3404: Long-running space reclamation task
**Version**
```
  "versions": [
    {
      "component": "core",
      "version": "3.21.0",
      "package": "pulpcore"
    },
    {
      "component": "container",
      "version": "2.14.2",
      "package": "pulp-container"
    },
    {
      "component": "rpm",
      "version": "3.18.5",
      "package": "pulp-rpm"
    },
    {
      "component": "python",
      "version": "3.7.2",
      "package": "pulp-python"
    },
    {
      "component": "ostree",
      "version": "2.0.0a6",
      "package": "pulp-ostree"
    },
    {
      "component": "file",
      "version": "1.11.1",
      "package": "pulp-file"
    },
    {
      "component": "deb",
      "version": "2.20.0",
      "package": "pulp_deb"
    },
    {
      "component": "certguard",
      "version": "1.5.5",
      "package": "pulp-certguard"
    },
    {
      "component": "ansible",
      "version": "0.15.0",
      "package": "pulp-ansible"
    }
  ],
```

Katello nightly (4.7)

**Describe the bug**

I noticed my reclaim space task was taking over 20 minutes in an environment with 63 repositories and 91485 rpm content units (to give some perspective).  PostgresSQL was being heavily taxed and taking 100% of one CPU core.  I tried to cancel it, but the cancellation was stuck so I needed to restart Pulpcore to stop the space reclamation.

Here's the task output after it was canceled forcefully:

```
{
  "pulp_href": "/pulp/api/v3/tasks/bce46114-a5d9-445a-a898-217210bf1975/",
  "pulp_created": "2022-11-15T16:38:50.639518Z",
  "state": "failed",
  "name": "pulpcore.app.tasks.reclaim_space.reclaim_space",
  "logging_cid": "c658f06c-3b76-49f6-a514-b19dd3bfbe52",
  "started_at": "2022-11-15T16:38:50.688113Z",
  "finished_at": "2022-11-15T17:09:06.918179Z",
  "error": {
    "reason": "Worker has gone missing."
  },
  "worker": "/pulp/api/v3/workers/80173b0a-f731-4c7b-b3ec-ed993369044e/",
  "parent_task": null,
  "child_tasks": [],
  "task_group": null,
  "progress_reports": [],
  "created_resources": [],
  "reserved_resources_record": [
    "shared:/pulp/api/v3/repositories/rpm/rpm/4ad1fb8e-ef06-42e6-a83a-00da97551dce/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/3224e11b-ec85-4e3d-8d7b-fd44dcfd184d/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/d0f49692-31dd-4709-9e52-27be83167a3f/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/bef78a95-9555-467b-9fe6-66650c081757/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/e5838919-ba35-4497-b8a0-98c10af8941b/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/7987e671-61e6-4d07-9c9b-ca7a07367d91/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/acd01e87-640a-4584-b52f-c999e937b55f/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/b01a1f40-c195-48c0-a05c-77b7748d6338/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/504b40fe-5d7f-456e-bc95-683878609791/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/8a1a3998-ff6c-460c-b26b-010ac57023a9/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/a1a44856-a028-4a2e-a539-aa73d3ef9ff3/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/1cde5855-eab1-4ac3-ac2f-f02a22541619/",
    "shared:/pulp/api/v3/repositories/deb/apt/509de38c-7ae7-4f7b-a37c-db8404488a51/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/cdd44804-8324-48ce-9e61-4ae6770d0427/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/dfe18547-f2bf-4c41-9b9e-32d6cb1e2f5e/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/d867837e-c35f-475d-9bb5-9c9bde465b19/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/a0bcd8d6-8e6d-4e05-83d1-8cbfbc28d8d9/",
    "shared:/pulp/api/v3/repositories/rpm/rpm/b0169f69-55cc-4ce1-830c-f444152c6853/"
  ]
}
```

**To Reproduce**
Run space reclamation on an environment with a similar amount of repositories and content to what I posted above.

**Expected behavior**
After chatting with @dralley it sounds like this may be slower performance than expected.

### PR DESCRIPTION
**Pull request by decko**
Closes #3404

### CODE REVIEW

**Review by dralley — APPROVED** [review]

File: CHANGES/3404.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Fixes a possible N+1 issue within reclaim_space task.
```
Comment (lubosmj) [inline CHANGES/3404.bugfix:1]: Could this be translated into a more user-friendly message with a past simple tense?

File: pulpcore/app/tasks/reclaim_space.py [hunk 2]
```diff
@@ -47,7 +47,7 @@ def reclaim_space(repo_pks, keeplist_rv_pks=None, force=False):
         if not content.cast().PROTECTED_FROM_RECLAIM:
             unprotected.append(content.pulp_type)
 
-    ca_qs = ContentArtifact.objects.filter(
+    ca_qs = ContentArtifact.objects.select_related("content").filter(
```
Comment (gerrod3) [inline pulpcore/app/tasks/reclaim_space.py:50]: Should this also include `"artifact"`? The loop below accesses the pk of the artifact. I know accessing `artifact_id` wouldn't trigger a db lookup, but I'm not sure if you access the pk off of the artifact object if that triggers a lookup as well.

Comment (dralley) [reply pulpcore/app/tasks/reclaim_space.py:50]: Good question - I *think* it's smart enough to avoid an extra query, maybe 90% sure, but not 100%

**Comment by mdellweg** [comment]
Am i right, that this change is not changing behaviour, but only performance?

**Comment by dralley** [comment]
@mdellweg  Yes

**Comment by mdellweg** [comment]
Then let's put the work "performance" in the changelog. ;)

File: pulpcore/app/tasks/reclaim_space.py [hunk 3]
```diff
@@ -47,8 +47,10 @@ def reclaim_space(repo_pks, keeplist_rv_pks=None, force=False):
...
 
-    ca_qs = ContentArtifact.objects.filter(
-        content__in=c_reclaim_qs.values("pk"), artifact__isnull=False
+    ca_qs = (
+        ContentArtifact.objects.select_related("content", "artifact")
+        .filter(content__in=c_reclaim_qs.values("pk"), artifact__isnull=False)
+        .values("content__pulp_type", "artifact__pk")
     )
     if not force:
         ca_qs = ca_qs.filter(remoteartifact__isnull=False)
```
Comment (dralley) [inline pulpcore/app/tasks/reclaim_space.py:56]: The error makes sense but tbh I wonder why it wasn't triggering before?

TL;DR since your queryset is using .values(), the objects returned are dictionaries rather than objects.  On line 60 you need to use something like `ca["content__pulp_type"]`

File: CHANGES/3404.bugfix (hunk 1 above)
Comment (decko) [reply CHANGES/3404.bugfix:1]: How could I write this in a better way? :thinking:

Comment (ggainey) [reply CHANGES/3404.bugfix:1]:
```suggestion
Addressed a possible N+1 performance issue within reclaim_space task.
```

wdyt?

File: CHANGES/3404.performance [hunk 4]
```diff
@@ -0,0 +1 @@
+Fixes a possible N+1 issue within reclaim_space task.
```
Comment (ggainey) [inline CHANGES/3404.performance:1]: We don't have a "performance" section of towncrier, so .bugfix it is. (Maybe we *should* have "performance" as an option - but that's a separate discussion/fix, elsewhere :) )

**Review by ggainey — CHANGES_REQUESTED** [review]

File: CHANGES/3404.performance (hunk 4 above)
Comment (decko) [reply CHANGES/3404.performance]: I was using `misc` in a previous iteration of this PR, but seems that a `bugfix` is applicable here. What do you think @mdellweg ?

Comment (ggainey) [reply CHANGES/3404.performance]: I like bugfix because it's good to let ppl know we're looking into performance issues. I believe .misc is "not visible to enduser" - and performance issues absolutely can be noticeable.

Comment (mdellweg) [reply CHANGES/3404.performance]: +1 To all of this. Sorry, i meant to use the word "performance" in the actual changelog line.

File: CHANGES/3404.bugfix [hunk 5]
```diff
@@ -0,0 +1 @@
+Addressed a possible N+1 performance issue within reclaim_space task.
```
Comment (mdellweg) [inline CHANGES/3404.bugfix:1]: I'm still a bit unsure about the "N+1" part. I assume you mean N+1 db calls instead of one. But at lease we tell our users that they can expect better perf with this fix and what was slow.

**Review by mdellweg — APPROVED** [review]

File: CHANGES/3404.bugfix (hunk 5 above)
Comment (decko) [reply CHANGES/3404.bugfix]: I changed it to "N+1 query performance issue" to be specific about the piece we worked on.

**Review by dralley — APPROVED** [review]

**Comment by dralley** [comment]
@ggainey Could you +1?

**Review by ggainey — APPROVED** [review]
